cpp/src/arrow/flight/client.cc

```diff
@@ -242,140 +244,239 @@ class GrpcClientAuthReader : public ClientAuthReader {
   }
 
  protected:
+  /// Get the gRPC status if not OK, to propagate any server error message
   Status OverrideWithServerError(Status&& st) {
-    // Get the gRPC status if not OK, to propagate any server error message
+    // TODO: this is a double-close when paired with GrpcStreamReader
     RETURN_NOT_OK(internal::FromGrpcStatus(stream_->Finish(), &rpc_->context));
     return std::move(st);
   }
 
  private:
-  GrpcStreamReader* flight_reader_;
   // The RPC context lifetime must be coupled to the ClientReader
   std::shared_ptr<ClientRpc> rpc_;
-  std::unique_ptr<grpc::ClientReader<pb::FlightData>> stream_;
+  std::shared_ptr<Reader> stream_;
+  std::shared_ptr<PeekableFlightDataReader<Reader>> peekable_reader_;
   bool stream_finished_;
 };
 
-GrpcStreamReader::GrpcStreamReader() {}
-
-Status GrpcStreamReader::Open(std::unique_ptr<ClientRpc> rpc,
-                              std::unique_ptr<grpc::ClientReader<pb::FlightData>> stream,
-                              std::unique_ptr<GrpcStreamReader>* out) {
-  *out = std::unique_ptr<GrpcStreamReader>(new GrpcStreamReader);
-  out->get()->rpc_ = std::move(rpc);
-  std::unique_ptr<GrpcIpcMessageReader> message_reader(
-      new GrpcIpcMessageReader(out->get(), out->get()->rpc_, std::move(stream)));
-  return ipc::RecordBatchStreamReader::Open(std::move(message_reader),
-                                            &(*out)->batch_reader_);
-}
+/// The implementation of the public-facing API for reading from a
+/// FlightData stream
+template <typename Reader>
+class GrpcStreamReader : public FlightStreamReader {
+ public:
+  GrpcStreamReader(std::shared_ptr<ClientRpc> rpc, std::shared_ptr<Reader> stream)
+      : rpc_(rpc),
+        stream_(stream),
+        peekable_reader_(new PeekableFlightDataReader<Reader>(stream)) {}
+
+  Status Open() override {
+    // TODO: this should internally be automatically called once we
+    // start seeing data in the stream
+    std::unique_ptr<GrpcIpcMessageReader<Reader>> message_reader(
+        new GrpcIpcMessageReader<Reader>(rpc_, stream_, peekable_reader_));
+    return ipc::RecordBatchStreamReader::Open(std::move(message_reader),
+                                              &this->batch_reader_);
+  }
 
-std::shared_ptr<Schema> GrpcStreamReader::schema() const {
-  return batch_reader_->schema();
-}
+  Status CheckStarted() const {
+    if (!batch_reader_) {
+      return Status::Invalid("Reader not initialized. Call Open().");
+    }
+    return Status::OK();
+  }
 
-Status GrpcStreamReader::Next(FlightStreamChunk* out) {
-  out->app_metadata = nullptr;
-  RETURN_NOT_OK(batch_reader_->ReadNext(&out->data));
-  out->app_metadata = std::move(last_app_metadata_);
-  return Status::OK();
-}
+  std::shared_ptr<Schema> schema() const override {
+    // TODO: this is no longer a simple getter
+    return batch_reader_->schema();
+  }
+  Status Next(FlightStreamChunk* out) override {
+    internal::FlightData* data;
+    peekable_reader_->Peek(&data);
```

Comment:
Could this fail (instead of the stream being finished)?

Reply:
See above - gRPC doesn't differentiate between a stream finishing and failing, unfortunately, on the client side.